infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -497,6 +436,79 @@ private void clearStreams(Set<UUID> streamsToClear) {
+            // no de-duplication on these updates will occur.  As a result, the transaction size can be large.
+            // Although it is within the maxApplySize limit, deserializing these entries to read the table can cause an
+            // OOM on applications running with a small memory footprint.  So for such tables, introduce an
+            // additional limit of max number of entries(50 by default) applied in a single transaction.  This
+            // algorithm is in line with the limits imposed in Compaction and Restore workflows.
+            if (bufferSize + smrEntry.getSerializedSize() + metadataSize > replicationContext.getConfig(session).getMaxApplySize()
+                    || maxEntriesLimitReached(streamId,
+                    buffer)) {
+                try (TxnContext txnContext = metadataManager.getTxnContext()) {
+                    updateLog(txnContext, buffer, streamId);
+                    if (!applyPhase) {
+                        ReplicationMetadata metadata = metadataManager.queryReplicationMetadata(txnContext, session);
+                        ReplicationMetadata updatedMetadata = metadata.toBuilder().setLastSnapshotStarted(srcGlobalSnapshot)
+                                .setLastSnapshotTransferredSeqNumber(currentSeqNum)
+                                .build();
+
+                        metadataManager.updateReplicationMetadata(txnContext, session, updatedMetadata);
+                    }
+                    ts = txnContext.commit();
+                    log.debug("Applied shadow stream partially for stream {} on address :: {}.  {} SMR entries written",
+                            streamId, ts.getSequence(), buffer.size());
+                    buffer.clear();
+                    buffer.add(smrEntry);
+                    bufferSize = smrEntry.getSerializedSize();
+                    numBatches++;
+                }
+            } else {
+                buffer.add(smrEntry);
+                bufferSize += smrEntry.getSerializedSize();
+            }
+        }
+        if (!buffer.isEmpty()) {
+            try (TxnContext txnContext = metadataManager.getTxnContext()) {
+                updateLog(txnContext, buffer, streamId);
+                if (!applyPhase) {
+                    ReplicationMetadata metadata = metadataManager.queryReplicationMetadata(txnContext, session);
+                    ReplicationMetadata updatedMetadata = metadata.toBuilder().setLastSnapshotStarted(srcGlobalSnapshot)
+                            .setLastSnapshotTransferredSeqNumber(currentSeqNum)
+                            .build();
+
+                    metadataManager.updateReplicationMetadata(txnContext, session, updatedMetadata);
+                }
+                ts = txnContext.commit();
+            }
+        }
+
+        if (snapshotSyncId != null && ts != null && !applyPhase && !snapshotSyncStartMarker.isPresent()) {
```

Comment:
We can move this into the empty buffer check block to avoid the null checks, can also have a wrapper method for both apply/transfer phase to simplify the usage since the snapshotSyncId and applyPhase params are not independent.

Reply:
Removed the null check of ts, and removed applyPhase as a param since as you said snapshotSyncId and applyPhase are not independent, we can use a applyPhase null check as applyPhase(like a boolean). However, it doesn't seem that we can move this block into the empty buffer check block, because we need the timestamp from there.  Instead, I moved all of the transfer phase specific code into 1 block to reduce the amount of if statements here.